crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2352,16 +2352,107 @@ impl<'db> TypeInferenceBuilder<'db> {
+            // iterators (crates, experimental APIs or manual impl might provide it)
+            .collect::<Vec<Type<'db>>>()
+            .windows(2)
+            .zip(ops.iter())
+            .enumerate()
+            .for_each(|(i, (pair, op))| {
+                let (left_ty, right_ty) = (pair[0], pair[1]);
+                if inferred_ty.is_none() {
+                    let is_last = i == ops.len() - 1;
+                    let comparison_ty = self.infer_binary_type_comparison(left_ty, *op, right_ty);
+                    // Special case for `Unknown` returned when an operation is not supported
+                    inferred_ty = if let Type::Unknown = comparison_ty {
+                        Some(Type::Unknown)
+                    } else {
+                        match (comparison_ty.bool(self.db), is_last) {
+                            (Truthiness::AlwaysTrue, false) => None,
+                            (Truthiness::AlwaysTrue, true) => Some(Type::BooleanLiteral(true)),
+                            (Truthiness::AlwaysFalse, _) => Some(Type::BooleanLiteral(false)),
+                            (Truthiness::Ambiguous, _) => {
+                                Some(builtins_symbol_ty(self.db, "bool").to_instance(self.db))
+                            }
+                        }
+                    };
+                }
+            });
+        inferred_ty.expect("A type should always be inferred on the last comparison in the chain")
+    }
+
+    /// Infers the type of a binary comparison (e.g. '<left> == <right>'). See
+    /// `infer_compare_expression` for the higher level logic dealing with multi-comparison
+    /// expressions.
+    fn infer_binary_type_comparison(
+        &mut self,
+        left: Type<'db>,
+        op: ast::CmpOp,
+        right: Type<'db>,
+    ) -> Type<'db> {
+        match (left, right) {
+            (Type::IntLiteral(n), Type::IntLiteral(m)) => Type::BooleanLiteral(match op {
+                ast::CmpOp::Eq | ast::CmpOp::Is => n == m,
+                ast::CmpOp::NotEq | ast::CmpOp::IsNot => n != m,
+                ast::CmpOp::Lt => n < m,
+                ast::CmpOp::LtE => n <= m,
+                ast::CmpOp::Gt => n > m,
+                ast::CmpOp::GtE => n >= m,
+                // Undefined for (int, int)
+                ast::CmpOp::In | ast::CmpOp::NotIn => false,
+            }),
+            // Booleans are coded as integers (False = 0, True = 1)
+            (Type::IntLiteral(n), Type::BooleanLiteral(b)) => self.infer_binary_type_comparison(
+                Type::IntLiteral(n),
+                op,
+                Type::IntLiteral(i64::from(b)),
+            ),
+            (Type::BooleanLiteral(b), Type::IntLiteral(m)) => self.infer_binary_type_comparison(
+                Type::IntLiteral(i64::from(b)),
+                op,
+                Type::IntLiteral(m),
+            ),
+            // TODO: move this logic wherever we handle class instances
```

Comment:
I don't think we need this TODO; we may end up consolidating some logic around e.g. treating `IntLiteral` as `int` when we can't handle it as a special case, but it's not clear if we will, and there's no missing or incomplete feature here we need to remember.

```suggestion
```